nodes/project-link.js

```diff
@@ -0,0 +1,847 @@
+                return null // side effect: undefined becomes null
+            } else if (value.type === 'Buffer') {
+                return Buffer.from(value.data)
+            } else if (value.type === 'bigint') {
+                return BigInt(value.data)
+            } else if (value.type === 'Map') {
+                return new Map(value.data)
+            } else if (value.type === 'Set') {
+                return new Set(value.data)
+            } else if (value.type === 'function') {
+                return new Function('return ' + value.data)()
+            }
+        }
+        return value
+    }
+
+    function parseLinkTopic (topic) {
+        // 0  1  2          3 4                                    5  6..
+        // ff/v1/7N152GxG2p/p/23d79df8-183c-4104-aa97-8915e1897326/in/a/b       pub proj→prog
+        // ff/v1/7N152GxG2p/p/ca65f5ed-aea0-4a10-ac9a-2086b6af6700/out/b1/b1    pub broadcast
+        // ff/v1/7N152GxG2p/p/23d79df8-183c-4104-aa97-8915e1897326/in/a/b       sub proj→prog
+        // ff/v1/7N152GxG2p/p/+/out/b1/b1                                       sub broadcast
+        const topicParts = (topic || '').split('/')
+        const projectOrDevice = topicParts[3] ? (topicParts[3] === 'd' ? 'd' : 'p') : null
+        const isBroadcast = topicParts[5] ? topicParts[5] === 'out' : null
+        // eslint-disable-next-line no-unused-vars
+        const isDirectTarget = topicParts[5] ? topicParts[5] === 'in' : null
+        const isCallResponse = topicParts[5] ? topicParts[5] === 'res' : null
+        const result = {
+            topicHeader: topicParts[0],
+            topicVersion: topicParts[1],
+            teamID: topicParts[2],
+            type: projectOrDevice,
+            projectID: topicParts[4],
+            isBroadcast,
+            isCallResponse,
+            subTopic: topicParts.slice(6).join('/')
+        }
+        return result
+    }
+
+    function buildLinkTopic (node, project, subTopic, isCallResponse) {
+        const topicParts = [TOPIC_HEADER, TOPIC_VERSION, RED.settings.flowforge.teamID]
+        if (node.type === 'project link in') {
+            topicParts.push('p')
+            if (project === 'broadcast') {
+                topicParts.push('+')
+                topicParts.push('out')
+            } else {
+                topicParts.push(project)
+                topicParts.push('in')
+            }
+        } else if (node.type === 'project link out') {
+            topicParts.push('p')
+            if (project === 'broadcast') {
+                topicParts.push(RED.settings.flowforge.projectID)
+                topicParts.push('out')
+            } else {
+                topicParts.push(project)
+                topicParts.push('in')
+            }
+        } else if (node.type === 'project link call') {
+            topicParts.push('p')
+            topicParts.push(project)
+            if (isCallResponse) {
+                topicParts.push('res')
```

Comment:
It looks like the node tries to use `ff/v1/<teamid>/p/<projectid>/res` - is that correct?

I haven't seen that topic mentioned in the design docs and it is *not* on the ACL list the platform provides - so this will not work with the current platform code.

We have two choices:

1. Use a special topic under the `/in` topic. We would need to ensure these messages don't get 'accidentally' received by a subscription to `/in/#`

2. Update the platform ACL to allow projects to subscribe to `..../res` 

Thoughts?

Reply:
Sorry, it was implied in the design but not mentioned (there was always a need to have a response to support the Link Call without having the user explicitly add a response path - i.e. keeping the project link nodes as simple and consistent). 

I initially had this as a mangled topic (e.g. '.../in/path_response`) but it meant substring type checks and felt ugly. 

So instead, simply went with an alternative sub path `/res/` - but unfortunately it didnt occur to me security would be that far down the path. Oopsy. My bad.

If possible, I would prefer the clean separation of path (as it is now) vs a mangled response topic that has to be filtered out as that has other implications.

Reply:
Can you add the ACLs?

We will need two adding - one for projects and one for devices. Here are the existing ACLs to allow them to subscribe to /in for reference.

 - allow project to subscribe to /in: https://github.com/flowforge/flowforge/blob/6589c285315245a5a0f21db3dd6c837d8d09b789/forge/ee/lib/projectComms/index.js#L17-L22
 - allow device to subscribe to /in: https://github.com/flowforge/flowforge/blob/6589c285315245a5a0f21db3dd6c837d8d09b789/forge/ee/lib/projectComms/index.js#L48-L54